File: apps/elixir/test/algorithm/search_algorithms_test.exs
defmodule Algorithm.SearchAlgorithmsTest do
  use ExUnit.Case
  alias Algorithm.SearchAlgorithms

  describe "linear_search/2" do
    test "値が見つかった場合にインデックスを返す" do
      assert SearchAlgorithms.linear_search([1, 3, 5, 7, 9], 5) == {:ok, 2}
    end

    test "値が見つからない場合に :not_found を返す" do
      assert SearchAlgorithms.linear_search([1, 3, 5], 4) == :not_found
    end
  end

  describe "binary_search/2" do
    test "ソート済みリストから値を二分探索する" do
      assert SearchAlgorithms.binary_search([1, 3, 5, 7, 9], 7) == {:ok, 3}
    end

    test "値が見つからない場合に :not_found を返す" do
      assert SearchAlgorithms.binary_search([1, 3, 5, 7, 9], 4) == :not_found
    end
  end

  describe "hash_search/2" do
    test "ハッシュマップから値を検索する" do
      map = %{1 => "a", 2 => "b", 3 => "c"}
      assert SearchAlgorithms.hash_search(map, 2) == {:ok, "b"}
      assert SearchAlgorithms.hash_search(map, 4) == :not_found
    end
  end
end

File: apps/scala/src/test/scala/algorithm/SearchAlgorithmsTest.scala
package algorithm

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class SearchAlgorithmsTest extends AnyFunSuite with Matchers:

  val arr = Array(1, 3, 5, 7, 9, 11, 13, 15)

  test("linearSearchWhile: キーが見つかる"):
    SearchAlgorithms.linearSearchWhile(arr, 7) shouldBe 3

  test("linearSearchWhile: キーが見つからない"):
    SearchAlgorithms.linearSearchWhile(arr, 6) shouldBe -1

  test("linearSearchFor: キーが見つかる"):
    SearchAlgorithms.linearSearchFor(arr, 5) shouldBe 2

  test("linearSearchFor: キーが見つからない"):
    SearchAlgorithms.linearSearchFor(arr, 2) shouldBe -1

  test("linearSearchSentinel: キーが見つかる"):
    SearchAlgorithms.linearSearchSentinel(arr, 9) shouldBe 4

  test("linearSearchSentinel: キーが見つからない"):
    SearchAlgorithms.linearSearchSentinel(arr, 4) shouldBe -1

  test("binarySearch: キーが見つかる"):
    SearchAlgorithms.binarySearch(arr, 11) shouldBe 5

  test("binarySearch: キーが見つからない"):
    SearchAlgorithms.binarySearch(arr, 10) shouldBe -1

  test("binarySearch: 先頭要素"):
    SearchAlgorithms.binarySearch(arr, 1) shouldBe 0

  test("binarySearch: 末尾要素"):
    SearchAlgorithms.binarySearch(arr, 15) shouldBe 7

  test("ChainedHash: 追加と検索"):
    val h = SearchAlgorithms.ChainedHash(13)
    h.add(1, "one") shouldBe true
    h.add(14, "fourteen") shouldBe true
    h.search(1) shouldBe Some("one")
    h.search(14) shouldBe Some("fourteen")
    h.search(999) shouldBe None

  test("ChainedHash: 重複追加は失敗"):
    val h = SearchAlgorithms.ChainedHash(13)
    h.add(1, "one") shouldBe true
    h.add(1, "one-dup") shouldBe false

  test("ChainedHash: 削除"):
    val h = SearchAlgorithms.ChainedHash(13)
    h.add(1, "one")
    h.remove(1) shouldBe true
    h.search(1) shouldBe None
    h.remove(1) shouldBe false

  test("OpenHash: 追加と検索"):
    val h = SearchAlgorithms.OpenHash(13)
    h.add(1, "one") shouldBe true
    h.search(1) shouldBe Some("one")
    h.search(999) shouldBe None

  test("OpenHash: 重複追加は失敗"):
    val h = SearchAlgorithms.OpenHash(13)
    h.add(1, "one") shouldBe true
    h.add(1, "one-dup") shouldBe false

  test("OpenHash: 削除"):
    val h = SearchAlgorithms.OpenHash(13)
    h.add(1, "one")
    h.remove(1) shouldBe true
    h.search(1) shouldBe None
    h.remove(1) shouldBe false
